### PR TITLE
[V0.0.8] Add GPU-aware placement (+ model knobs and logging API changes)

### LINKED ISSUE #73: Rework model logging
Discussion @ https://issues.apache.org/jira/browse/SINGA-412

### LINKED ISSUE #75: Improve API for configuring model knobs
Discussion @ https://issues.apache.org/jira/browse/SINGA-413

### LINKED ISSUE #78: Allow preferential placement of train workers on nodes with GPUs

### PR DESCRIPTION
**Pull request by nginyc**
Fixes #73 
Fixes #75 
Fixes #78 

Consolidates changes from branches `improve_knob_api` and `improve_model_logging` reviewed by @nudles.

The new docs for this PR are available at https://nginyc.github.io/rafiki/docs/0.0.8.

GPU-aware placement: 
- Uses [Docker Swarm placement constraints](https://docs.docker.com/engine/swarm/services/#placement-constraints)
- https://nginyc.github.io/rafiki/docs/0.0.8/docs/src/dev/setup.html#adding-nodes-to-rafiki
- https://github.com/nginyc/rafiki/blob/add_gpu_aware_placement/scripts/setup_node.sh

Model knobs configuration changes:
- https://nginyc.github.io/rafiki/docs/0.0.8/docs/src/user/creating-models.html#implementing-models
- https://nginyc.github.io/rafiki/docs/0.0.8/docs/src/python/rafiki.model.html#knob-types
-https://github.com/nginyc/rafiki/blob/add_gpu_aware_placement/examples/models/image_classification/TfFeedForward.py

Model logging API changes:
- https://nginyc.github.io/rafiki/docs/0.0.8/docs/src/user/creating-models.html#logging-in-models
- https://nginyc.github.io/rafiki/docs/0.0.8/docs/src/python/rafiki.model.html#rafiki.model.ModelLogger
- https://github.com/nginyc/rafiki/blob/add_gpu_aware_placement/examples/models/image_classification/TfFeedForward.py

### CODE REVIEW

File: examples/models/image_classification/TfFeedForward.py [hunk 1]
```diff
@@ -8,110 +8,89 @@
...
-        dataset = self.utils.load_dataset_of_image_files(dataset_uri, image_size=[self._image_size, self._image_size])
+        im_sz = self._knobs.get('image_size')
+        bs = self._knobs.get('batch_size')
+        ep = self._knobs.get('epochs')
+
+        logger.log('Available devices: {}'.format(str(device_lib.list_local_devices())))
+
+        # Define 2 plots: Loss against time, loss against epochs
+        logger.define_loss_plot()
+        logger.define_plot('Loss Over Time', ['loss'])
```
Comment (nudles) [inline examples/models/image_classification/TfFeedForward.py]: timestamp is always recorded.
can we always plot 2 figures including one with x-axis being the timestamp?

File: examples/models/image_classification/TfFeedForward.py [hunk 2]
```diff
@@ -8,110 +8,89 @@
...
                     callbacks=[
                         tf.keras.callbacks.LambdaCallback(on_epoch_end=self._on_train_epoch_end)
                     ]
                 )
 
                 # Compute train accuracy
                 (loss, accuracy) = self._model.evaluate(images, classes)
-                self.utils.log('Train loss: {}'.format(loss))
-                self.utils.log('Train accuracy: {}'.format(accuracy))
+                logger.log('Train loss: {}'.format(loss))
```
Comment (nudles) [inline examples/models/image_classification/TfFeedForward.py]: how about
`logger.log(loss=loss)`

File: examples/models/image_classification/TfFeedForward.py [hunk 3]
```diff
@@ -8,110 +8,89 @@
...
 
-    def init(self, knobs):
-        self._batch_size = knobs.get('batch_size')
-        self._hidden_layer_units = knobs.get('hidden_layer_units')
-        self._hidden_layer_count = knobs.get('hidden_layer_count')
-        self._learning_rate = knobs.get('learning_rate')
-        self._epochs = knobs.get('epochs')
-        self._image_size = knobs.get('image_size')
-
+    def __init__(self, **knobs):
```
Comment (nudles) [inline examples/models/image_classification/TfFeedForward.py]: use `self.__dict__.update(knobs)` for one example.

File: rafiki/model/log.py [hunk 4]
```diff
@@ -1,73 +1,190 @@
...
+        from rafiki.model import logger, BaseModel
+        ...
+        class MyModel(BaseModel):
+            ...
+            def train(self, dataset_uri):
+                ...
+                logger.log('Starting model training...')
+                logger.define_plot('Precision & Recall', y_axis=['precision', 'recall'])
+                ...
+                logger.log(precision=0.1, recall=0.6, epoch=1)
```
Comment (nudles) [inline rafiki/model/log.py]: need to put epoch in `logger.define_plot('Precision & Recall', y_axis=['precision', 'recall'], x_axis='epoch')`

**Comment by nginyc** [comment]
Made minor changes based on the feedback

**Comment by nginyc** [comment]
Was approved by @nudles
